OpenRA.Mods.AS/Projectiles/ElectricBolt.cs

```diff
@@ -117,7 +117,7 @@ public ElectricBolt(ElectricBoltInfo info, ProjectileArgs args)
 					upVector = 1024 * upVector / upVector.Length;
 			}
 
-			zaps = new HashSet<(Color, WPos[])>();
+			zaps = new List<(Color Color, WPos[] Positions, WPos[] PosCache)>();
```

Comment:
Is the HashSet->List change really required here? Hashset is better for perf.

Reply:
The place in this new modification which use `zaps` is:
1. initialize at line 120 and 121
2. access all its elements by using iterator (`foreach`) at line 139

The `HashSet` has the major advantage that any specific element inside it can be accessed immediately, at the cost of memory usage.
But line 139 we are using `foreach`, which (should be optimized) access one by one by using a iterator.

As for iterator, I cannot explain weill on this term, maybe this [link](https://stackoverflow.com/questions/16327461/what-is-the-difference-between-iteration-and-traversing) is close to what I am thinking.

Reply:
and use iterator of `List` to access  `List`  should be fast on traverse all element (see List data struct)

Reply:
@Mustapha-TR I don't know if my understanding is correct on C# on `List` (I use my C++ knowlege on it but data struct should not has big change right?)

Reply:
Actually, you cannot apply C++ assumptions to C# data types (assumptions like such leads to PRs like https://github.com/OpenRA/OpenRA/pull/17267).

Most of the C# data types implement various primitive interfaces (IEnumerable, IList and such) and it is these primitive interfaces and their handling are what defines an object type's capabilities.

It is true that HashSet has memory disadvantage here compared to List. But we don't care about the ordering here and we only use it as an iterator and the memory constant isn't a big deal either IMO because these projectiles have a short lifetime and garbage-collected regularly. Last I checked, iterating a hashset was faster than iterating a list and this was important to decrease pressure during rendering. Sure, one can argue that with this getting moved to tick, I am worrying on overoptimization.